### PR TITLE
Stop writing to column `user_id` of tables `profiles` and `user_filters`

### PR DESCRIPTION
**Pull request by H-Shay**
This is part 4 of a migration to drop the column `user_id` from the tables `profiles` and `user_filters`. In this stage we cease writing to the column in preparation for dropping it entirely in part 5. 

In order to continue writing to the other columns of the tables while ceasing to write to `user_id`, the `NOT NULL` constraint on this column is dropped as well.

### CODE REVIEW

File: synapse/storage/database.py [hunk 1]
```diff
@@ -98,6 +98,8 @@
     "event_push_summary": "event_push_summary_unique_index2",
     "receipts_linearized": "receipts_linearized_unique_index",
     "receipts_graph": "receipts_graph_unique_index",
+    "profiles": "profiles_full_user_id_key_idx",
+    "user_filters": "full_users_filters_unique_idx",
```
Comment (MatMaul) [inline synapse/storage/database.py:102]: FTR, this is included in this PR because it is the one that begin using `full_user_id` for the upserts.

**Review by MatMaul — APPROVED** [review]
LGTM.

**Comment by H-Shay** [comment]
@MatMaul thanks for reviewing - this had been sitting for awhile and after your approval I double-checked it and realized it was subject to the same error I fixed in https://github.com/matrix-org/synapse/pull/15817. I have fixed the code here, would you mind double-checking that it still looks sane? Thanks!

File: synapse/storage/schema/main/delta/79/01_drop_user_id_constraint_profiles.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        # irritatingly in SQLite we need to rewrite the table to drop the constraint.
+        cur.execute("DROP TABLE IF EXISTS temp_profiles")
+
+        create_sql = """
+        CREATE TABLE temp_profiles (
+            full_user_id text NOT NULL,
+            user_id text,
+            displayname text,
+            avatar_url text,
+            UNIQUE (full_user_id),
```
Comment (MatMaul) [inline synapse/storage/schema/main/delta/79/01_drop_user_id_constraint_profiles.py:26]: Regarding your last comment @H-Shay , same pattern here I believe ?

**Review by MatMaul — CHANGES_REQUESTED** [review]

File: synapse/storage/schema/main/delta/79/01_drop_user_id_constraint_profiles.py (hunk 2 above)
Comment (H-Shay) [reply synapse/storage/schema/main/delta/79/01_drop_user_id_constraint_profiles.py:26]: The `profiles` table does want these rows to be `UNIQUE` - this should be correct.

**Comment by H-Shay** [comment]
Merged #15787 into develop, thanks for the (double) reviews @MatMaul!

**Pull request merged**